Segments sorted by non-time columns.

gianm opened the pull request:
## Summary

Currently, segments are always sorted by `__time`, followed by the sort order provided by the user via `dimensionsSpec` or `CLUSTERED BY`. Sorting by `__time` enables efficient execution of queries involving time-ordering or granularity. Time-ordering is a simple matter of reading the rows in stored order, and granular cursors can be generated in streaming fashion.

However, for various workloads, it's better for storage footprint and query performance to sort by arbitrary orders that do not start with `__time`. With this patch, users can sort segments by such orders.

## API

For spec-based ingestion, users add `forceSegmentSortByTime: false` to dimensionsSpec. The `dimensions` list determines the sort order. To define a sort order that includes `__time`, users explicitly include a dimension named `__time`.

For SQL-based ingestion, users set the context parameter `forceSegmentSortByTime: false`. The CLUSTERED BY clause is then used as the explicit segment sort order.

In both cases, when the new `forceSegmentSortByTime` parameter is `true` (the default), `__time` is implicitly prepended to the sort order, as it always was prior to this patch.

The new parameter is experimental for two main reasons. First, such segments can cause errors when loaded by older servers, due to violating their expectations that timestamps are always monotonically increasing. Second, even on newer servers, not all queries can run on non-time-sorted segments. Scan queries involving time-ordering and any query involving granularity will not run. (To partially mitigate this, a currently-undocumented SQL feature `sqlUseGranularity` is provided. When set to false the SQL planner avoids using `granularity`.)

## Main changes

Changes on the write path:

1) `DimensionsSpec` can now optionally contain a `__time` dimension, which controls the placement of `__time` in the sort order. If not present, `__time` is considered to be first in the sort order, as it has always been.

2) `IncrementalIndex` and `IndexMerger` are updated to sort facts more flexibly; not always by time first.

3) `Metadata` (stored in metadata.drd) gains a `sortOrder` field.

4) MSQ can generate range-based shard specs even when not all columns are singly-valued strings. It merely stops accepting new clustering key fields when it encounters the first one that isn't a singly-valued string. This is useful because it enables range shard specs on `someDim` to be created for clauses like `CLUSTERED BY someDim, __time`.

5) Auto-compaction respects and propagates sort orders that don't start with `__time`.

Changes on the read path:

1) Update cursor holders for `QueryableIndex` and `IncrementalIndex` to return the ordering of the underlying index, so query engines can tell how a segment is sorted.

2) Update `CursorGranularizer` and `VectorCursorGranularizer` to throw errors when using granularities on non-time-ordered segments.

3) Update `ScanQueryEngine` to throw an error when using the time-ordering `order` parameter on non-time-ordered segments.

4) Update `TimeBoundaryQueryRunnerFactory` to perform a segment scan when running on a non-time-ordered segment.

5) Add `sqlUseGranularity` context parameter that causes the SQL planner to avoid using granularities other than ALL. This is undocumented, because it's hopefully a short-term hack. The more ideal thing would be to have all the native queries work properly on these segments.

6) Move `getMinTime`, `getMaxTime`, and `getMaxIngestedEventTime` from `StorageAdapter` to `TimeBoundaryInspector` and `MaxIngestedEventTimeInspector`. This is mainly necessary for `timeBoundary` queries to be able to tell whether they can use `getMinTime` and `getMaxTime`, vs. needing to use a cursor. Previously, `timeBoundary` assumed that an adapter backing a `TableDataSource` was guaranteed to have an exact `getMinTime` and `getMaxTime`. But after this patch, that isn't necessarily going to be the case (in particular: a non-time-sorted segment won't be able to know its exact min/max time without a full scan.)

Other changes:

1) Rename `DimensionsSpec#hasCustomDimensions` to `hasFixedDimensions` and change the meaning subtly: it now returns true if the `DimensionsSpec` represents an unchanging list of dimensions, or false if there is some discovery happening. This is what call sites had expected anyway.

2) Removed `descending` from `Joinable#makeJoinMatcher`. Now that `descending` has more or less been phased out in favor of `ordering: [__time DESC]`, the joinable order no longer needs to be reversed. (All joined rows arising from the same left-hand row have the same value.)

gianm commented on processing/src/main/java/org/apache/druid/segment/TimeAndDimsPointer.java:
```diff
@@ -148,19 +177,23 @@
+      } else {
+        final int dimIndex = index < timePosition ? index : index - 1;
+        final int dimDiff = dimensionSelectorComparators[dimIndex].compare(
+            dimensionSelectors[dimIndex],
```
It can't be out of bounds, because if it's easy to the array length, it would get decremented for being `>= timePosition`.

gianm commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -887,19 +942,23 @@
-      if (!dimensionDescs.isEmpty()) {
-        throw new ISE("Cannot load dimension order when existing order[%s] is not empty.", dimensionDescs.keySet());
+      if (size() != 0) {
+        throw new ISE("Cannot load dimension order when existing index is not empty.", dimensionDescs.keySet());
```
fixed.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/Metadata.java:
```diff
@@ -233,6 +260,39 @@ public String toString()
+    }
+
+    if (sortOrdersToMerge.stream().anyMatch(Objects::isNull)) {
+      return null;
```
should this be an error if they aren't all null? seems like it should be pretty consistent across indexable adapters...

gianm replied:
I made it lenient for two reasons:

- Merging of the other parts of the Metadata is also lenient
- It isn't documented that this method requires that all Metadatas are sourced from segments created with the same ingestion spec + the same version of the software

clintropolis commented on processing/src/main/java/org/apache/druid/segment/Metadata.java:
```diff
@@ -233,6 +260,39 @@ public String toString()
+      String column = null;
+
+      for (final List<String> sortOrder : sortOrdersToMerge) {
+        if (mergedSortOrder.size() >= sortOrder.size()) {
```
does this happen from columns with all null values or something? maybe this method could use some comments to make the rationale behind the decisions clearer

gianm replied:
I added some comments and also fixed a problem: `null` is now treated as `[__time]` (which makes sense, since that was the only possible sort order prior to the `sortOrder` field being added). Due to this change the method is no longer `@Nullable`.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/RowBasedStorageAdapter.java:
```diff
@@ -167,6 +168,15 @@ public Metadata getMetadata()
+  @Override
+  public List<String> getSortOrder()
+  {
+    // It's possibly incorrect in some cases for sort order to be SORTED_BY_TIME_ONLY here, but for historical reasons,
+    // we're keeping this in place for now. The handling of "interval" in "makeCursors", which has been in place for
+    // some time, suggests we think the data is always sorted by time.
```
is this referring to the `RowWalker` with its `skipToDateTime`?

gianm replied:
Yes.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -334,6 +349,17 @@ protected IncrementalIndex(
+        incrementalIndexSchema.getTimestampSpec(),
+        this.gran,
+        this.rollup,
+        ColumnHolder.TIME_COLUMN_NAME.equals(Iterables.getFirst(dimensionOrder, null)) ? null : dimensionOrder
```
maybe we should always write this out instead of leaving it null? i guess it is done like this to make it easier to fill in older segments and new segments ordered by time first?

gianm replied:
Hmm I don't totally remember why this is conditional. I updated it to always write out.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndexStorageAdapter.java:
```diff
@@ -295,6 +300,16 @@ public Metadata getMetadata()
+    if (index.timePosition == 0) {
+      return Metadata.SORTED_BY_TIME_ONLY;
+    } else {
+      return Collections.emptyList();
```
i suppose depending on the type of facts holder we could actually report something here (rollup should be ordered i think?)

gianm replied:
I think that makes sense. I added a comment about this being a possible change in the future.

gianm commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -777,11 +790,18 @@ public Iterator<IncrementalIndexRow> iterator(boolean descending)
+        ConcurrentMap<IncrementalIndexRow, IncrementalIndexRow> rangeMap = descending ? subMap.descendingMap() : subMap;
+        return rangeMap.keySet();
+      } else {
+        return Iterables.filter(
```
I think it's also used for `intervals` filtering even for non-granular cursors. It seems like it'd be useful for that at least.

gianm commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -918,4 +940,84 @@ public void clear()
+    }
+
+    @Override
+    public Iterator<IncrementalIndexRow> iterator(boolean descending)
```
It seemed to make more sense to implement it here rather than throw an error. I suppose in theory there could be a use case, in the future, for situations like `ORDER BY userId DESC` when the segment is sorted by `userId`.

gianm commented on processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:
```diff
@@ -256,6 +256,23 @@ public Metadata getMetadata()
+  {
+    final List<String> baseSortOrder = baseAdapter.getSortOrder();
+
+    // Sorted the same way as the base segment, unless the unnested column shadows one of the base columns.
```
I figure the common case is that the unnested column will not be first in the sort order. So we'll return whatever prefix of the order is there up to the unnested column.

gianm commented on processing/src/main/java/org/apache/druid/segment/StorageAdapter.java:
```diff
@@ -129,6 +141,18 @@ default RowSignature getRowSignature()
+   * this ordering, using the natural comparator for the type of the column as returned by
+   * {@link #getColumnCapabilities(String)}. Includes {@link ColumnHolder#TIME_COLUMN_NAME} if appropriate.
+   */
+  List<String> getSortOrder();
+
+  default boolean isTimeOrdered()
+  {
+    return ColumnHolder.TIME_COLUMN_NAME.equals(Iterables.getFirst(getSortOrder(), null));
+  }
```
This sounds like a good idea, although in the interests of minimizing conflicts, it'd probably be good to do this after #16533 is merged (since it moves `OrderBy` around).

gianm commented on processing/src/main/java/org/apache/druid/segment/Metadata.java:
```diff
@@ -100,6 +112,14 @@ public Boolean isRollup()
+  @Nullable
+  @JsonProperty
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public List<String> getSortOrder()
+  {
+    return sortOrder;
+  }
+
```
see https://github.com/apache/druid/pull/16849#discussion_r1708084545

gianm commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:
```diff
@@ -620,10 +620,18 @@ public static Function<Set<DataSegment>, Set<DataSegment>> addCompactionStateToS
+      if (ingestionSpec.getDataSchema().getDimensionsSpec() == null) {
+        dimensionsSpec = null;
+      } else {
+        // We do not need to store dimensionExclusions and spatialDimensions since auto compaction does not support them
```
will update.

gianm commented:
From the discussion in https://github.com/apache/druid/pull/16849#discussion_r1708084545, after #16533 is merged we should revisit this one to (a) resolve conflict (b) replace `List<String>` with `List<OrderBy>` in the sort order field.

gianm commented:
Pushed up a commit resolving conflicts with #16533. Main changes:

- Changed frame cursors to report "no ordering" rather than time ordering.
- Removed `descending` from `Joinable#makeJoinMatcher`. Now that `descending` has more or less been phased out in favor of `ordering: [__time DESC]`, the joinable order no longer needs to be reversed. (All joined rows arising from the same left-hand row have the same value.)
- Moved getMinTime, getMaxTime, and getMaxIngestedEventTime from StorageAdapter to TimeBoundaryInspector and MaxIngestedEventTimeInspector. This is mainly necessary for timeBoundary queries to be able to tell whether they can use getMinTime and getMaxTime, vs. needing to use a cursor. Previously, timeBoundary assumed that an adapter backing a TableDataSource was guaranteed to have an exact getMinTime and getMaxTime. But after this patch, that isn't necessarily going to be the case (in particular: a non-time-sorted segment won't be able to know its exact min/max time without a full scan.)

clintropolis reviewed (approved):
love it, this is exciting :metal:

gianm commented on docs/querying/sql-query-context.md:
```diff
@@ -50,11 +50,11 @@ Configure Druid SQL query planning using the parameters in the table below.
+|`maxNumericInFilters`|Max limit for the amount of numeric values that can be compared for a string type dimension when the entire SQL WHERE clause of a query translates only to an [OR](../querying/filters.md#or) of [Bound filter](../querying/filters.md#bound-filter). By default, Druid does not restrict the amount of of numeric Bound Filters on String columns, although this situation may block ot ...
+|`inFunctionThreshold`| At or beyond this threshold number of values, SQL `IN` is converted to [`SCALAR_IN_ARRAY`](sql-functions.md#scalar_in_array). A threshold of 0 forces this conversion in all cases. A threshold of [Integer.MAX_VALUE] disables this conversion. The converted function is eligible for fewer planning-time optimizations, which speeds up planning, but may prevent certain planning-t ...
+|`inFunctionExprThreshold`|At or beyond this threshold number of values, SQL `IN` is eligible for execution using the native function `scalar_in_array` rather than an <code>&#124;&#124;</code> of `==`, even if the number of values is below `inFunctionThreshold`. This property only affects translation of SQL `IN` to a [native expression](math-expr.md). It does not affect translation of SQL `IN` to ...
+|`inSubQueryThreshold`|At or beyond this threshold number of values, SQL `IN` is converted to `JOIN` on an inline table. `inFunctionThreshold` takes priority over this setting. A threshold of 0 forces usage of an inline table in all cases where the size of a SQL `IN` is larger than `inFunctionThreshold`. A threshold of `2147483647` disables the rewrite of SQL `IN` to `JOIN`. |`2147483647`|
```
I meant to leave it undocumented, for the reason mentioned in the PR description,

> This is undocumented, because it's hopefully a short-term hack. The more ideal thing would be to have all the native queries work properly on these segments.

clintropolis replied:
oops, forgot about that :+1:

gianm commented:
Pushed a change to make the parameter `forceSegmentSortByTime` (default `true`) rather than `useExplicitSegmentSortOrder` (default `false`).

I think this makes more sense, because it refers to "time" explicitly (this is really about time) and it also would work better for a migration to changing the default. If the default changes to not force sort by time, then users wanting backwards compatibility could set `forceSegmentSortByTime: true`.

vogievetsky commented on docs/ingestion/ingestion-spec.md:
```diff
@@ -232,7 +232,7 @@ A `dimensionsSpec` can have the following components:
 | `includeAllDimensions` | Note that this field only applies to string-based schema discovery where Druid ingests dimensions it discovers as strings. This is different from schema auto-discovery where Druid infers the type for data. You can set `includeAllDimensions` to true to ingest both explicit dimensions in the `dimensions` field and other dimensions that the ingestion task discovers from in ...
 | `useSchemaDiscovery` | Configure Druid to use schema auto-discovery to discover some or all of the dimensions and types for your data. For any dimensions that aren't a uniform type, Druid ingests them as JSON. You can use this for native batch or streaming ingestion.  | false  | 
-
+| `forceSegmentSortByTime` | When set to true (the default), segments created by the ingestion job are sorted by `{__time, dimensions[0], dimensions[1], ...}`. When set to false, segments created by the ingestion job are sorted by `{dimensions[0], dimensions[1], ...}`. To include `__time` in the sort order when this parameter is set to `false`, you must include a dimension named `__time` with typ ...
```
You need to update the default (3rd column) to `true`

vogievetsky reviewed (approved):
Looks good from an API perspective. Expect a followup web console PR soon.

maytasm commented:
@clintropolis @gianm 
I think this PR broke dataSourceMetadata query.
Seems like MaxIngestedEventTimeInspector in DataSourceMetadataQueryRunner will always be null.
This can be reproduced with the wikipedia dataset.
Just run
```
{
"queryType" : "dataSourceMetadata",
"dataSource" : "wikipedia"
}
```
The "maxIngestedEventTime" returned is -146136543-09-08T08:23:32.096Z is wrong.

Doesn't look like this code:
```
  @SuppressWarnings("unchecked")
  @Nullable
  @Override
  public <T> T as(@Nonnull Class<T> clazz)
  {
    final Function<QueryableIndexSegment, ?> fn = AS_MAP.get(clazz);
    if (fn != null) {
      final T fnApply = (T) fn.apply(this);
      if (fnApply != null) {
        return fnApply;
      }
    }

    if (TimeBoundaryInspector.class.equals(clazz)) {
      return (T) timeBoundaryInspector;
    } else if (Metadata.class.equals(clazz)) {
      return (T) index.getMetadata();
    } else if (PhysicalSegmentInspector.class.equals(clazz)) {
      return (T) new QueryableIndexPhysicalSegmentInspector(index);
    } else if (TopNOptimizationInspector.class.equals(clazz)) {
      return (T) new SimpleTopNOptimizationInspector(true);
    }

    return Segment.super.as(clazz);
  }
  ```
Is handling MaxIngestedEventTimeInspector?

clintropolis commented:
>I think this PR broke dataSourceMetadata query.
Seems like MaxIngestedEventTimeInspector in DataSourceMetadataQueryRunner will always be null.

The intention of the new `MaxIngestedEventTimeInspector` is that it is only for use with realtime data, because every other implementation on the old `StorageAdapter` of `getMaxIngestedEventTime()` was just calling `getMaxTime()` which gives the end interval of the segment. Could use `timeBoundary` instead if there is no realtime data.

It does seem odd we return min datetime, i would have expected null. I don't think it makes much sense to have an implementation of `MaxIngestedEventTimeInspector` for published segments given that `timeBoundary` exists, but the docs could probably more clearly spell this out. Apologies for the behavior change.

maytasm commented:
@clintropolis 
Thanks for the response.
- Im a little confused on when you said that getMaxTime() gives the end interval of the segment. I believe that this query (dataSourceMetadata) was previously returning something like `2025-01-21T08:23:32.096Z` for our non-realtime datasource. The code for getMaxTime() in QueryableIndexStorageAdapter looks something like:
```
  private void populateMinMaxTime()
  {
    // Compute and cache minTime, maxTime.
    final ColumnHolder columnHolder = index.getColumnHolder(ColumnHolder.TIME_COLUMN_NAME);
    try (NumericColumn column = (NumericColumn) columnHolder.getColumn()) {
      this.minTime = DateTimes.utc(column.getLongSingleValueRow(0));
      this.maxTime = DateTimes.utc(column.getLongSingleValueRow(column.length() - 1));
    }
  }
```
which does seems to return the actual time (min and max) of the rows/data (not the end interval of the segment)... unless I am missing something.
- Given that the intention of the new MaxIngestedEventTimeInspector is for only realtime data, is it still fair to say that the dataSourceMetadata query should still work for non-realtime datasource (as it was previously working for non-realtime datasources)?
- It is returning min datetime as currMaxIngestedEventTime is null at: https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/datasourcemetadata/DataSourceMetadataQuery.java#L109

maytasm commented:
- Ah, the timeBoundary does return the actual time (min and max) of the rows/data which matches with the previous dataSourceMetadata. We can switch to the timeBoundary query as a workaround. Thanks!

clintropolis commented:
>Im a little confused on when you said that getMaxTime() gives the end interval of the segment. 

Oh sorry, i meant the max timestamp within the segment interval, that code is the same, its just on `TimeBoundaryInspector` now, https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/QueryableIndexTimeBoundaryInspector.java#L85

gianm commented:
Oops, the behavior change was unintentional (at least I didn't intend it, and this was my PR). This should restore the old behavior: #17686
